core/src/test/scala/unit/kafka/log/LogCleanerTest.scala

```diff
@@ -1451,6 +1451,60 @@ class LogCleanerTest {
     log.close()
   }
 
+  @Test
+  def testRecoveryRebuildsIndices(): Unit = {
```

Comment:
Would be good to enumerate different cases that we could run into during recovery and ensure we have coverage for them. eg.

1. All `.swap` files are present. We should validate that producer state is not rebuilt in this case.
2. Some `.swap` and some `.clean` files are present.
3. All `.clean` files are present.
4. One of the index files was not found when completing the swap operation, which triggers a full recovery and rebuild of producer state.